src/mac.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/mac.h"
+#include "mococrw/error.h"
+#include "mococrw/openssl_wrap.h"
+
+#include <openssl/hmac.h>
+
+namespace mococrw
+{
+
+MessageAuthenticationCode::~MessageAuthenticationCode() = default;
+
+class HMAC::Impl
+{
+public:
+    Impl(openssl::DigestTypes hashFunction, const std::vector<uint8_t> key, size_t outputLength) :
+        _outputLength(outputLength)
+    {
+        const EVP_MD* digestFn = _getMDPtrFromDigestType(hashFunction);
+
+        if (key.empty()) {
+            throw MoCOCrWException("Key for HMAC is empty.");
+        }
+
+        _ctx = openssl::_HMAC_CTX_new();
+        openssl::_HMAC_Init_ex(_ctx.get(), key.data(), key.size(), digestFn, NULL);
+    }
+
+    ~Impl() = default;
```

Comment:
Secure clearing of key memory?

Reply:
This should be done by the std::unique_ptr<HMAC_CTX>:
using SSL_HMAC_CTX_Ptr = std::unique_ptr<HMAC_CTX, SSLDeleter<HMAC_CTX, lib::OpenSSLLib::SSL_HMAC_CTX_free>>;

This is invoking OPENSSL_cleanse(ctx->key, sizeof(ctx->key));
As we don't store the key anywhere in between that should be save.

Reply:
ah nice